File: Week 7/2.PalindromeAndAnagram.java
/*
 * WEEK 7 - STRINGS
 * Topic: Palindrome Check & Anagram Check
 *
 * PALINDROME: A string that reads the same forwards and backwards.
 * Examples: "racecar", "madam", "level", "A man a plan a canal Panama"
 *
 * APPROACH: Two pointers — one from start, one from end, move inward.
 * Time: O(n), Space: O(1)
 *
 * ANAGRAM: Two strings are anagrams if one is a rearrangement of the other.
 * Examples: "listen" and "silent", "eat" and "tea"
 *
 * APPROACH 1: Sort both strings and compare. O(n log n), O(1) extra.
 * APPROACH 2: Frequency count array. O(n), O(1) space (fixed 26-element array).
 * APPROACH 3: HashMap for unicode/general characters. O(n), O(k) space (k = unique chars).
 */

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public class PalindromeAndAnagram {

    // --- PALINDROME ---

    // Simple palindrome check (two-pointer)
    static boolean isPalindrome(String s) {
        int left = 0, right = s.length() - 1;
        while (left < right) {
            if (s.charAt(left) != s.charAt(right)) return false;
            left++;
            right--;
        }
        return true;
    }

    // Palindrome ignoring non-alphanumeric characters and case
    // Example: "A man, a plan, a canal: Panama" → true
    static boolean isPalindromeIgnoreNonAlpha(String s) {
        int left = 0, right = s.length() - 1;
        while (left < right) {
            // Skip non-alphanumeric from left
            while (left < right && !Character.isLetterOrDigit(s.charAt(left))) left++;
            // Skip non-alphanumeric from right
            while (left < right && !Character.isLetterOrDigit(s.charAt(right))) right--;
            if (Character.toLowerCase(s.charAt(left)) != Character.toLowerCase(s.charAt(right)))
                return false;
            left++;
            right--;
        }
        return true;
    }

    // --- ANAGRAM ---

    // Approach 1: Sort and compare — O(n log n)
    static boolean isAnagramSort(String a, String b) {
        if (a.length() != b.length()) return false;
        char[] ca = a.toCharArray(), cb = b.toCharArray();
        Arrays.sort(ca);
        Arrays.sort(cb);
        return Arrays.equals(ca, cb);
    }

    // Approach 2: Frequency count — O(n), O(1) space (26-char array)
    static boolean isAnagramFreq(String a, String b) {
        if (a.length() != b.length()) return false;
        int[] freq = new int[26];
        for (char c : a.toCharArray()) freq[c - 'a']++;
        for (char c : b.toCharArray()) {
            freq[c - 'a']--;
            if (freq[c - 'a'] < 0) return false;
        }
        return true;
    }

    // Approach 3: HashMap — works for any characters including unicode
    static boolean isAnagramMap(String a, String b) {
        if (a.length() != b.length()) return false;
        Map<Character, Integer> map = new HashMap<>();
        for (char c : a.toCharArray()) map.merge(c, 1, Integer::sum);
        for (char c : b.toCharArray()) {
            if (!map.containsKey(c) || map.get(c) == 0) return false;
            map.merge(c, -1, Integer::sum);
        }
        return true;
    }

    public static void main(String[] args) {
        // --- Palindrome tests ---
        System.out.println("=== Palindrome ===");
        String[] tests = {"racecar", "hello", "level", "madam", "a", ""};
        for (String t : tests) {
            System.out.printf("isPalindrome(\"%s\") = %b%n", t, isPalindrome(t));
        }

        System.out.println("\nisPalindromeIgnoreNonAlpha:");
        System.out.println("\"A man, a plan, a canal: Panama\" = " +
            isPalindromeIgnoreNonAlpha("A man, a plan, a canal: Panama")); // true
        System.out.println("\"race a car\" = " +
            isPalindromeIgnoreNonAlpha("race a car")); // false

        // --- Anagram tests ---
        System.out.println("\n=== Anagram ===");
        String[][] pairs = {{"listen", "silent"}, {"eat", "tea"}, {"hello", "world"}, {"anagram", "nagaram"}};
        for (String[] p : pairs) {
            System.out.printf("\"%s\" vs \"%s\":%n", p[0], p[1]);
            System.out.println("  Sort:  " + isAnagramSort(p[0], p[1]));
            System.out.println("  Freq:  " + isAnagramFreq(p[0], p[1]));
            System.out.println("  Map:   " + isAnagramMap(p[0], p[1]));
        }
    }
}

File: Week 7/rust/2.palindrome_and_anagram.rs
/*
 * WEEK 7 - RUST DSA
 * Topic: Palindrome & Anagram
 * File: 2.palindrome_and_anagram.rs
 *
 * CONCEPT:
 *     Palindrome: same forwards and backwards.
 *     Anagram   : same characters with same multiplicities.
 *
 * KEY POINTS:
 *     - For ASCII inputs the .as_bytes() view enables O(1) random access.
 *     - For arbitrary UTF-8, materialise a Vec<char> first.
 *     - Frequency-array (26) for ASCII letters; HashMap for any chars.
 *
 * ALGORITHM / APPROACH:
 *     Palindrome (two-pointer over bytes):
 *         let bytes = s.as_bytes();
 *         let (mut l, mut r) = (0usize, bytes.len() - 1);
 *         while l < r {
 *             if bytes[l] != bytes[r] { return false; }
 *             l += 1; r -= 1;
 *         }
 *         true
 *
 * RUST-SPECIFIC NOTES:
 *     - char::is_alphanumeric is Unicode-aware; for ASCII use is_ascii_alphanumeric.
 *     - sort a Vec<char> with .sort() then compare slices.
 *     - Use std::collections::HashMap for the map approach.
 *
 * DRY RUN:
 *     "racecar" -> palindrome (true)
 *     "race a car" ignore non-alpha -> false
 *     "listen"/"silent" sorted both = "eilnst" -> anagram
 *
 * COMPLEXITY:
 *     Palindrome  : O(n)
 *     Sort anagram: O(n log n)
 *     Freq anagram: O(n)
 *     Map  anagram: O(n)
 */

use std::collections::HashMap;

fn is_palindrome(s: &str) -> bool {
    let bytes = s.as_bytes();
    if bytes.is_empty() {
        return true;
    }
    let mut l: usize = 0;
    let mut r: usize = bytes.len() - 1;
    while l < r {
        if bytes[l] != bytes[r] {
            return false;
        }
        l += 1;
        r -= 1;
    }
    true
}

fn is_palindrome_ignore_non_alpha(s: &str) -> bool {
    let chars: Vec<char> = s.chars().collect();
    if chars.is_empty() {
        return true;
    }
    let mut l: i64 = 0;
    let mut r: i64 = chars.len() as i64 - 1;
    while l < r {
        while l < r && !chars[l as usize].is_alphanumeric() {
            l += 1;
        }
        while l < r && !chars[r as usize].is_alphanumeric() {
            r -= 1;
        }
        let a = chars[l as usize].to_ascii_lowercase();
        let b = chars[r as usize].to_ascii_lowercase();
        if a != b {
            return false;
        }
        l += 1;
        r -= 1;
    }
    true
}

fn is_anagram_sort(a: &str, b: &str) -> bool {
    if a.len() != b.len() {
        return false;
    }
    let mut va: Vec<char> = a.chars().collect();
    let mut vb: Vec<char> = b.chars().collect();
    va.sort();
    vb.sort();
    va == vb
}

fn is_anagram_freq(a: &str, b: &str) -> bool {
    if a.len() != b.len() {
        return false;
    }
    let mut freq = [0i32; 26];
    for c in a.bytes() {
        freq[(c - b'a') as usize] += 1;
    }
    for c in b.bytes() {
        let i = (c - b'a') as usize;
        freq[i] -= 1;
        if freq[i] < 0 {
            return false;
        }
    }
    true
}

fn is_anagram_map(a: &str, b: &str) -> bool {
    if a.chars().count() != b.chars().count() {
        return false;
    }
    let mut m: HashMap<char, i32> = HashMap::new();
    for c in a.chars() {
        *m.entry(c).or_insert(0) += 1;
    }
    for c in b.chars() {
        let entry = m.entry(c).or_insert(0);
        if *entry == 0 {
            return false;
        }
        *entry -= 1;
    }
    true
}

fn main() {
    println!("=== Palindrome ===");
    for t in ["racecar", "hello", "level", "madam", "a", ""] {
        println!("is_palindrome(\"{}\") = {}", t, is_palindrome(t));
    }

    println!("\nis_palindrome_ignore_non_alpha:");
    let s1 = "A man, a plan, a canal: Panama";
    println!("\"{}\" = {}", s1, is_palindrome_ignore_non_alpha(s1));
    let s2 = "race a car";
    println!("\"{}\" = {}", s2, is_palindrome_ignore_non_alpha(s2));

    println!("\n=== Anagram ===");
    let pairs = [
        ("listen", "silent"),
        ("eat", "tea"),
        ("hello", "world"),
        ("anagram", "nagaram"),
    ];
    for (a, b) in pairs {
        println!("\"{}\" vs \"{}\":", a, b);
        println!("  Sort: {}", is_anagram_sort(a, b));
        println!("  Freq: {}", is_anagram_freq(a, b));
        println!("  Map:  {}", is_anagram_map(a, b));
    }
}

/*
 * NOTES — Rust vs Java:
 *     - .as_bytes() gives O(1) ASCII indexing; .chars() iterates UTF-8 code points.
 *     - HashMap::entry().or_insert(...) replaces map.merge() pattern.
 *     - char::is_alphanumeric is Unicode-aware; use is_ascii_* for ASCII-only.
 *     - We use i64 for the alphanumeric two-pointer to avoid usize underflow.
 */
